Fix of #3240 (Drop Temporary file usage while reading data)

rizac opened the pull request:
### What does this PR do?

This PR tries to fix the first issue (point 1) in #3240, allowing now to seamlessly read from both file-like objects and file paths and eliminating the highly inefficient fallback of writing data to disk on `TypeError`s (if these errors happen now, there is not need for the fallback, and `read` simply raises them)

Note for reviewers:

1. (fixed, see thread on this page) ~~I could not write support for file-like objects in the case of 'GCF' format because the code in the related module was quite criptic: as such, this format still allows only file paths as argument~~ 
2. The point 2. mentioned in #3240, i.e. removing temporary files when reading from URL, requires more workaround and I dropped it for the moment

In account of the two points above, I am available to improve my PR or provide an additional one, but probably some discussion is needed. I am available even via zoom in case as the problem might require some round table and pair programming

### Why was it initiated?  Any relevant Issues?

See point 1. of #3240 

### PR Checklist
- [x] Correct base branch selected? master for new features, maintenance_... for bug fixes
- [ ] This PR is not directly related to an existing issue (which has no PR yet).
- [ ] While the PR is still work-in-progress, the no_ci label can be added to skip CI builds
- [ ] If the PR is making changes to documentation, docs pages can be built automatically.
  Just add the build_docs tag to this PR.
  Docs will be served at [docs.obspy.org/pr/{branch_name}](https://docs.obspy.org/pr/) (do not use master branch).
Please post a link to the relevant piece of documentation.
- [ ] If all tests including network modules (e.g. clients.fdsn) should be tested for the PR,
   just add the test_network tag to this PR.
- [x] All tests still pass (locally).
- [ ] Any new features or fixed regressions are covered via new tests.
- [ ] Any new or changed features are fully documented.
- [ ] Significant changes have been added to CHANGELOG.txt .
- [ ] First time contributors have added your name to CONTRIBUTORS.txt .
- [ ] If the changes affect any plotting functions you have checked that the plots
   from all the CI builds look correct. Add the "upload_plots" tag so that plotting
   outputs are attached as artifacts.
- [ ] New modules, add the module to CODEOWNERS with your github handle
- [ ] Add the yellow ready for review label when you are ready for the PR to be reviewed.

paitor commented:
Hi

Don't have any objections to the changes done to the gcf reader but perhaps that the added guard in lines 214 and 217 seems a bit superfluous, but perhaps I missing something here.

Could you please give some more details what was cryptic in the code base and I can perhaps walk you through this.

rizac commented:
> Hi
> 
> Don't have any objections to the changes done to the gcf reader but perhaps that the added guard in lines 214 and 217 seems a bit superfluous, but perhaps I missing something here.
> 
> Could you please give some more details what was cryptic in the code base and I can perhaps walk you through this.

Hi 
thanks for the reply.

The problem with gcf is that I do not see a line where the file is open, so I cannot work out a solution where I replace that `open` with a file-like object, which I basically did seamlessly for all other cases. 
If you look at the code for `_is_gcf` (`read_gcf` below in the module works the same):

https://github.com/obspy/obspy/blob/417b047a2c13491232a3b472a9cf3c0a78785f22/obspy/io/gcf/core.py#L210-L212

You will see that `gcf_io.read_gcf` accepts a file **path** (encoded as bytes seq.) and the `_GcfFile` object. I do not see any way here to let the function work with a file like object

Thank you very much for any solution
(Ps: I am not an expert of the format so keep in mind I might miss something trivial for you)

paitor commented:
Hi

First of all, thanks for working on this think it'll be a nice contrib.

The python part of the code sends the file path to the underlying c-code where the file is open, see: 

https://github.com/obspy/obspy/blob/417b047a2c13491232a3b472a9cf3c0a78785f22/obspy/io/gcf/src/gcf_io.c#L534-L540

where `open_r` and `ORFLAG` are macros defined in:

 https://github.com/obspy/obspy/blob/417b047a2c13491232a3b472a9cf3c0a78785f22/obspy/io/gcf/src/gcf_io.h

in order to have the c-code compile on (most) platforms. So what needs to be done to the code i:
1.  to move the opening of the file from the C-code to the python code (i.e. into `read_gcf`)
2. figure out how to convert the binary stream into a proper file descriptor to send to the c-code
3. in the c-code remove the call to `opengcf`. 
I've never done 2. so can't really advice you how to go about with this (but perhaps have a look in the other readers that have c-code under the hood, e.g. mseed).

Also in `_is_gcf` there is a check that the input is a file and then the size of the file as a gcf file always must be an integer multiple of 1024 bytes. This is a fairly quick first test to discard files that does not fit the criteria hence I think it would be nice to keep it.  It's not known to me though how to extract this information from a from any type of file-like object but I assume that this can be done.

rizac commented:
Thanks @paitor for the suggestion. 

I slightly changed your proposed solution to keep away from the rabbit hole of file descriptors, C and O/S compatibility (which I tried to sort out without success). 

I therefore proceed to:

1. Made the gcf C implementation more nodular, and therefore `read_gcf` easier to translate into Python
3. Copied that implementation into pure Python, and from there manage the file and file-like objects more easily

Tests are passing on my local machine (btw, why are they not passing on Github CLI?), let me know if additional work is needed

rizac commented:
@patior Good point, no I did not run any performance test, let's see what the other reviewers say.

However, in the framework of the current PR, we should always benchmark any performance change with the performance improvements of this PR, also in terms of avoiding to implement custom code to prevent writing several files to disk needlessly. 

I am trying to refactor text files because I saw some improvement that could be done to the code, then I would also wait for other reviewers because the PR is quite needs feedback from many people I guess

rizac commented:
Any feedback?

megies commented:
Sorry for not seeing this earlier. I'll give this a proper look and review soon! :grimacing:

megies commented on obspy/io/gcf/core.py:
```diff
@@ -372,19 +452,27 @@ def _read_gcf(filename, headonly=False, network='', station='',
+
+    if isinstance(file, IOBase):
+        if ret:
+            raise IOError("failed to read in-memory GCF data (error code %d)"
```
if it's `IOBase`, isn't it "in-memory" misleading? Looks more like a file handle, potentially or in general?

megies commented on obspy/io/gse2/core.py:
```diff
@@ -6,27 +6,27 @@
-            libgse2.is_gse2(f)
-    except Exception:
-        return False
+    with open_bytes_stream(file) as stream:
```
In general we exclusively use that variable name "stream" only for `Stream` objects, but it's pretty clear what is up here, so I dont think we need to be too picky here, just as a side note

rizac replied:
Changed all occurrences of `as stream:` into `as io_stream:` (3 in total, not a big deal)

megies commented on obspy/io/gse2/core.py:
```diff
@@ -107,27 +107,25 @@ def _write_gse2(stream, filename, inplace=False, **kwargs):  # @UnusedVariable
-        return True
-    return False
+        return data.startswith(b'WID1') \
+            or data.startswith(b'XW01')
```
This return statement can be dedented, no reason to have it inside the `with` block

megies commented on obspy/io/mseed/tests/test_mseed_reading_and_writing.py:
```diff
@@ -155,6 +155,7 @@ def test_downcast_int64_to_int32(self):
         # make sure the data can be written to mseed
         with io.BytesIO() as buf:
             st.write(buf, format="mseed")
+            buf.seek(0, 0)
```
I don't understand how this worked before without the `seek()`, huh

rizac replied:
honestly, I thought the same when writing that line

megies commented:
Oh and this might need a rebase eventually too

rizac commented:
Thanks @megies , I'll try to go through all your comments the next days and commit changes to this PR
(hope there isn't any particular forthcoming deadline)

megies commented:
> (hope there isn't any particular forthcoming deadline)

It's in the github milestone for 1.5.0 and it won't get left out for sure, no worries

rizac commented on obspy/io/gcf/core.py:
```diff
@@ -180,49 +181,138 @@ def merge_gcf_stream(st):
         obj = _GcfFile()
-        b_filename = filename.encode('utf-8')
-        ret = gcf_io.read_gcf(b_filename, obj, 3)
+        ret = _fill_gcf_from_file(file, obj, 3)
```
Considering that I am a little rusty on C and have no expertise on Gcf formats, as you said the priority was to change as little as possible rather than being Pythonic (for instance, that function is intended to mimic the C counterpart) and also make life easier for reviewers when comparing with old implementation (see discussion with @paitor above, but in general if you guys know other potential experts on the subject please invite them to get their feedback)

rizac commented on obspy/io/alsep/pse/tape.py:
```diff
@@ -27,9 +27,29 @@ def __enter__(self):
+            # (see self.fromfile)
+            self._handle.fileno()
+            self._handle_has_fileno = True
+        except Exception:
```
except io.UnsupportedOperation

rizac commented on obspy/io/alsep/wt/tape.py:
```diff
@@ -20,18 +21,32 @@ def __enter__(self):
+            # (see self.fromfile)
+            self._handle.fileno()
+            self._handle_has_fileno = True
+        except Exception:
```
except io.UnsupportedOperation

rizac commented on obspy/io/alsep/core.py:
```diff
@@ -16,16 +17,29 @@
+        # objects. In all other cases, use np.frombuffer:
+        try:
+            return np.fromfile(file, dtype='u1', shape=16)
+        except Exception:
```
this was actually a bug, thanks @megies : `shape` is a typo and should be `count`. Tests were passing because of the bare Except.
Fixed now with `count` instead of `shape` and `UnsupportedOperation` instead of bare Except

rizac commented on obspy/io/ascii/core.py:
```diff
@@ -72,7 +71,7 @@ def _is_slist(filename):
     """
     try:
-        with open(filename, 'rt') as f:
+        with open_text_stream(filename, encoding='ascii') as f:
```
If the package is ASCII, and files are supposed to be ASCII, then the format should be "ascii". Not changing this for the moment, unless I miss something, e.g. in time series ascii files "ascii" is some sort of legacy but they do mean text files encoded in various formats, in that case I will remove it

rizac commented on obspy/io/ascii/core.py:
```diff
@@ -130,11 +129,12 @@ def _read_slist(filename, headonly=False, **kwargs):  # @UnusedVariable
         buf = []
         key = False
         for line in fh:
+            line = line
```
(harmless) typo

rizac commented:
Fixed the last comments of @megies (thanks for the review), apart from the ascii question (see comment to the only unresolved issue above).

As a side note, just for safety, as @megies pointed out **this PR touches the very core of things** and nobody of us is expert in everything, if you guys know any person experienced in specific code blocks that can check the modifications (or run this branch in their usual workflow) to see if there are additional tests to add, I would be happy to have them invited to the discussion

megies commented:
I rebased on current master and force pushed to get fresh CI results

megies commented:
@rizac there are some problems with GCF on C level during the waveform plugin tests.

`$ obspy-runtests -v -k test_waveform_plugins.py -k test_is_format`

This goes into a segmentation fault. I narrowed it down a bit to basically this, running `_is_gcf()` on two cybershake test files back to back:

```python
from obspy.io.gcf.core import _is_gcf

_is_gcf("obspy/obspy/io/cybershake/tests/data/test.grm")
_is_gcf("obspy/obspy/io/cybershake/tests/data/not_cybershake.grm")
```

```
free(): invalid next size (fast)
Aborted
```

I'll leave the rest to you.

rizac commented:
@megies  I'll have a look asap. Hope this is solvable easily.
I am just puzzled why I did not have this error beforehand. Do I have everything to reproduce the bug? Were some new tests added (just for info)?

paitor commented:
@rizac 

I tested megies code snippet with my initial PR of the gcf reader but could not reproduce. Therefore I had a quick look at the current PR but can't see an obvious cause of this. However, I note an issue (already present in my initial PR) that may be related which you may want to fix while you are at it.

On lines 222 and 224 the function simply returns False meaning that the release of memory in the `finally` section of the `try` clause will not be reached, potentially causing a memory leakage. Better then to introduce a variable `retval = True` at the beginning of the function, change lines 222 and 224 to set this to `False` and at the end of the function return this variable.

That said this causes another issue in that the releasing of memory by `gcf_io.free_GcfFile` in line 228 may end up calling `free` on uninitialized memory. To avoid this, add after line 212 the lines:
```
    gcf_io.init_GcfFile.argtypes = [ctypes.POINTER(_GcfFile)]
    gcf_io.init_GcfFile.restype = None
```
and then directly after `obj = _GcfFile()` add the line
```
        gcf_io.init_GcfFile(obj)
```
which will make sure pointers in `obj` are properly initialized to `NULL`

Hope this makes sense and with some luck it will resolve the issue (although then it seems strange that I failed to reproduce)

rizac commented:
Hi @megies @paitor et al,

I commit and rebased (hope I did everything correctly, not a master in rebasing). Let's cross our fingers. Please note that:

1. I could not reproduce the bug either, so I basically followed step by step @paitor suggestions (thanks a lot). 
@paitor **please check if I implemented all your suggestions correctly**: your line refs (assuming they are related to [gcf/core.py](https://github.com/rizac/obspy/commit/f390a3e94277a0cd8451eb1f74bc82ab3674777f#diff-5982d423eab952ddf93199c22a43aa25b5d4b8b3b8ce163374d2f70b17f4594c)) are shifted by 10 on my editor and the link above (why?) so I did not  find my way around easily

2. Tests are now failing locally, but errors arise from `io/cybershake` which I did not modify and seems unrelated to this fix (maybe some merge happened in the meantime?). Let me know if I have to care about it, but I might need some help in case

paitor commented:
@rizac Looks good to me and thanks for fixing my errors (I'm still not all friend with github, probably due to laziness)

megies commented:
@rizac for me this is still failing like I reported last week, see https://github.com/obspy/obspy/pull/3346#issuecomment-2109935612

Did you reinstall obspy from source (i.e. recompile C parts) after switching to your PR's branch (e.g. `pip uninstall -y obspy && pip install -e ~/git/obspy`)?

megies commented:
Are you on Mac? Maybe that's why you can't reproduce, it seems OK on Mac.

> 2\. Tests are now failing locally, but errors arise from `io/cybershake` which I did not modify and seem unrelated to this fix (maybe some merge happened in the meantime?). Let me know if I have to care about it, but I might need some help in case

All fails I see relate to `_is_gcf`, if you look at the tracebacks even if the fails are in `io.alsep` or `io.cybershake` it is the auto-format detection calling `_is_gcf()` causing issues, see e.g. https://tests.obspy.org/139563/#6

As far as I can see, **fails on windows** are due to a missing symbol in `gcf/src/gcf_io.def` (see above test report) and then **Mac seems OK** and there are **no test reports for Linux** getting sent to our test server for the other platforms since `_is_gcf()` is crashing in C and pytest crashes as well before sending a report.

Note that there is a different error if you comment out the first line:

```python
from obspy.io.gcf.core import _is_gcf

_is_gcf("/home/megies/git/obspy/obspy/io/cybershake/tests/data/test.grm")
_is_gcf("/home/megies/git/obspy/obspy/io/cybershake/tests/data/not_cybershake.grm")
```

```
corrupted size vs. prev_size
Aborted
```

and this is different:


```python
from obspy.io.gcf.core import _is_gcf

#_is_gcf("/home/megies/git/obspy/obspy/io/cybershake/tests/data/test.grm")
_is_gcf("/home/megies/git/obspy/obspy/io/cybershake/tests/data/not_cybershake.grm")
```

```
free(): invalid next size (fast)
Aborted
```

Also note that the file which surfaces the issue has very simple short content but that should be no issue for the code:

```
with open("obspy/io/cybershake/tests/data/not_cybershake.grm", "rb") as fh:
    data = fh.read()
```

`b'0\n'`

rizac commented:
Ok it looks like the problem arised when the gcf block is not a multiple of 1024, now fixed, returning -1 (L 298 of io/gcf/core.py). Still getting errors when testing locally but apparently unrelated.
Let me know, hope everything it's fixed on my side

@megies I did not rebase for the moment because I might need other fixes, just added a new commit. Also, yes, I am on Mac, I can test on Ubuntu, but no Windows

megies reviewed (approved):
Alright, tests look OK now, let's merge and finger's crossed :rofl:
